File: backend/src/main/java/com/dreampath/service/chatbot/RagAnswerService.java
package com.dreampath.service.chatbot;

import lombok.RequiredArgsConstructor;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class RagAnswerService {

    @Value("${openai.api.key}")
    private String openaiApiKey;

    private final OkHttpClient client = new OkHttpClient();

    public String generateAnswer(String question, JSONArray matches) {

        StringBuilder context = new StringBuilder();

        for (int i = 0; i < matches.length(); i++) {
            JSONObject m = matches.getJSONObject(i);

            if (m.has("metadata")) {
                JSONObject metadata = m.getJSONObject("metadata");

                if (metadata.has("question") && metadata.has("answer")) {
                    String faqQ = metadata.getString("question");
                    String faqA = metadata.getString("answer");
                    context.append("Q: ").append(faqQ).append("\n");
                    context.append("A: ").append(faqA).append("\n\n");
                } else if (metadata.has("text")) {
                    context.append("- ").append(metadata.getString("text")).append("\n");
                }
            }
        }

        String prompt = """
                당신은 FAQ 기반 고객 상담 챗봇입니다.
                아래 FAQ 정보를 참고하여 사용자의 질문에 답변해주세요.

                중요:
                - FAQ의 질문과 사용자 질문이 유사한 의미라면 해당 FAQ의 답변을 사용하세요.
                - FAQ 답변을 자연스럽게 다시 말해서 전달하세요.
                - FAQ에 관련 정보가 전혀 없을 때만 "죄송하지만 해당 질문에 대한 정보가 없습니다. 메인페이지 하단 문의하기를 통해 문의를 남겨주세요."라고 답변하세요.

                FAQ 정보:
                %s

                사용자 질문: %s

                친절하고 자연스럽게 한국어로 답변하세요.
                """.formatted(context.toString(), question);

        try {

            JSONObject msg = new JSONObject()
                    .put("role", "user")
                    .put("content", prompt);

            JSONObject json = new JSONObject()
                    .put("model", "gpt-4o-mini")
                    .put("messages", new JSONArray().put(msg));

            RequestBody body = RequestBody.create(
                    json.toString(),
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url("https://api.openai.com/v1/chat/completions")
                    .addHeader("Authorization", "Bearer " + openaiApiKey)
                    .post(body)
                    .build();

            Response response = client.newCall(request).execute();
            String responseBody = response.body().string();

            JSONObject obj = new JSONObject(responseBody);
            return obj.getJSONArray("choices")
                    .getJSONObject(0)
                    .getJSONObject("message")
                    .getString("content");

        } catch (Exception e) {
            throw new RuntimeException("GPT 답변 생성 실패", e);
        }
    }
}

File: ai-service/services/chatbot/rag_answer_service.py
import os
from openai import OpenAI
from typing import List, Dict, Any


class RagAnswerService:
    def __init__(self):
        self.client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

    def generate_answer(self, question: str, matches: List[Dict[str, Any]]) -> str:
        """FAQ 전용 답변 생성 메서드"""

        # FAQ 형식으로 context 생성
        context = []

        for match in matches:
            metadata = match.get("metadata", {})

            if "question" in metadata and "answer" in metadata:
                faq_q = metadata["question"]
                faq_a = metadata["answer"]
                context.append(f"Q: {faq_q}\nA: {faq_a}\n")
            elif "text" in metadata:
                context.append(f"- {metadata['text']}\n")

        # FAQ가 없으면 무조건 범위 외 메시지 반환
        if not context or len(context) == 0:
            return "죄송하지만, DreamPath 서비스 관련 질문 외에는 답변할 수 없습니다. DreamPath와 관련된 질문이나 도움이 필요하시면 언제든지 말씀해 주세요!😊"

        context_str = "\n".join(context)

        # FAQ가 있으면 GPT로 자연스러운 답변 생성
        prompt = f"""당신은 DreamPath 진로 상담 서비스의 친절한 AI 어시스턴트입니다.

아래 참고 정보를 바탕으로 사용자의 질문에 친절하고 자연스럽게 한국어로 답변하세요.

**중요 규칙:**
1. 사용자 질문이 참고 정보의 내용과 관련이 있다면, 참고 정보를 바탕으로 친절하게 답변하세요.
2. 사용자 질문이 DreamPath와 전혀 관련 없는 일상적인 질문(예: 날씨, 음식 추천, 일반 상식 등)이라면,
   다음 메시지를 **정확히 그대로** 반환하세요:
   "죄송하지만, DreamPath 서비스 관련 질문 외에는 답변할 수 없습니다. DreamPath와 관련된 질문이나 도움이 필요하시면 언제든지 말씀해 주세요! 😊"

참고 정보:
{context_str}

사용자 질문: {question}

답변:"""

        try:
            response = self.client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )
            return response.choices[0].message.content

        except Exception as e:
            raise RuntimeError(f"GPT 답변 생성 실패: {str(e)}")
